### DISCUSSION TITLE
Compiling IfcConvert for use on iOS and Swift (dylib)

### DISCUSSION THREAD

**Discussion by jusssch** [IfcOpenShell/IfcOpenShell#3846] [Q&A]
Hey, i am currently building an ifc viewer on mobile (iOS to be specific) and it is going good so far. But now i have ran into some issues. See, when i want to open a ifc file from iOS (Swift) i need to convert the file to something like dae or another file format so that i can render it with SceneKit. The problem is that i really have problems with compiling. I get errors like the following:

`/Users/myname/IfcOpenShell-0.7.0/src/ifcconvert/../serializers/../ifcgeom_schema_agnostic/../ifcparse/../ifcparse/IfcParse.h:157:32: error: definition or redeclaration of 'string' not allowed inside a function static my_thread_local std::string s;`

or 

`/Users/myname/IfcOpenShell-0.7.0/src/ifcconvert/../serializers/../ifcgeom_schema_agnostic/../ifcparse/IfcFile.h:166:19: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions] bool batch_mode_ = false;`

and many more... 

This is really stopping me. I haven't found something on the internet for this issue. Also even if i compile to a dylib (which is necessary because of the license of IfcOpenShell) i don't know how to call the functions of IfcConvert (e.g. create a Swift Binding for the library). 

Help would be really much appreciated. I really don't know how to continue from here...

-Julius

**Comment by aothms** [comment]
>     ... C++11 extension ...

It's likely that your not compiling with the correct C++ specification version. If you use CMake (which you can use to generate the XCode project files) you get the correct configuration (which is C++14) https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/cmake/CMakeLists.txt#L21-L22

> Also even if i compile to a dylib (which is necessary because of the license of IfcOpenShell) i don't know how to call the functions of IfcConvert (e.g. create a Swift Binding for the library).

I agree this is likely the right approach to deal with LGPL license compliance. Language interoperability like this is not necessarily trivial though and requires a fairly good understanding of the compilation and linking process. The most straight forward way possible is to define your own C (so not C++) interface to the library functionality you use and compile that into a DLL. C doesn't use name mangling because it doesn't support polymorphism. See a guide like this https://www.swiftprogrammer.info/swift_call_cpp.html

I would also recommend at least investigating another approach of doing the model conversion online. That way you're not distributing LGPL code (because it doesn't leave the server) so you're not really bound by license complexity. And you also don't have to fight with fairly complex language interoperability issues. There are likely cases anyway where an IFC model is so large and complex that it can't be processed on mobile hardware.

**Reply by jusssch** [reply]
Hi, thank you for your answer! 

I use CMake. How would i switch to get the correct configuration (which as you said is C++14)? 

> See a guide like this https://www.swiftprogrammer.info/swift_call_cpp.html

That guide helped understanding how it generally works. That is good but it did not really help to use on the IfcOpenShell library as i can not compile it without knowing how to compile with C++14. Also with a use case like IfcConvert those files would be extremely big? I have in mind to drastically delete files i don't need. Which ones could i even delete?

Let's think about the process the user goes through. He has downloaded a ifc file and wants to open it with the app through the file picker. After clicking on the file (if the user has no internet) the app has the file in memory but how would i then pass this file into the IFCConvert function and more importantly how would i get the file back?

> I would also recommend at least investigating another approach of doing the model conversion online.

Converting of ifc files to other files online is also in planning but i really haven't found an application for Apple platforms at least in which i can open a ifc file directly on mobile offline. If possible i will try to do the conversion somewhere on a server but if not i really would like to stick with the idea of processing it locally.

I would be really happy about an answer!
Thank you in advance.

-Julius

**Reply by aothms** [reply]
> That is good but it did not really help to use on the IfcOpenShell library as i can not compile it without knowing how to compile with C++14

It ought to work automatically. That's the 'joy' of using cmake. See https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html which is what we use. 

What do you see under the C++ language dialect then in the XCode build settings?

> the app has the file in memory but how would i then pass this file into the IFCConvert function and more importantly how would i get the file back?

Well, that's indeed one of the things to think through. You probably also want to do some caching so that if a user opens the same file again it would instantly load. It doesn't necessarily need to be file based though. The IfcGeom::Iterator will give you bunch of triangulated data to visualize (which in IfcConvert the serializers will write to disk).
